Core/HLE/HLE.cpp

```diff
@@ -354,6 +403,136 @@ bool hleIsKernelMode() {
+	args.resize(argc);
+	memcpy(args.data(), argv, argc * sizeof(u32));
+
+	enqueuedMipsCalls.push_back({ func, afterAction, args });
+
+	hleAfterSyscall |= HLE_AFTER_QUEUED_CALLS;
+}
+
+void hleFlushCalls() {
+	u32 &sp = currentMIPS->r[MIPS_REG_SP];
+	PSPPointer<HLEMipsCallStack> stackData;
+	VERBOSE_LOG(HLE, "Flushing %d HLE mips calls from %s, sp=%08x", (int)enqueuedMipsCalls.size(), latestSyscall ? latestSyscall->name : "?", sp);
+
+	// First, we'll add a marker for the final return.
+	sp -= sizeof(HLEMipsCallStack);
+	stackData.ptr = sp;
+	stackData->nextOff = 0xFFFFFFFF;
+	stackData->ra = currentMIPS->pc;
+	stackData->v0 = currentMIPS->r[MIPS_REG_V0];
+	stackData->v1 = currentMIPS->r[MIPS_REG_V1];
+
+	// Now we'll set up the first in the chain.
+	currentMIPS->pc = enqueuedMipsCalls[0].func;
+	currentMIPS->r[MIPS_REG_RA] = HLEMipsCallReturnAddress();
+	for (int i = 0; i < (int)enqueuedMipsCalls[0].args.size(); i++) {
+		currentMIPS->r[MIPS_REG_A0 + i] = enqueuedMipsCalls[0].args[i];
+	}
+
+	// For stack info, process the first enqueued call last, so we run it first.
+	// We don't actually need to store 0's args, but keep it consistent.
+	for (int i = (int)enqueuedMipsCalls.size() - 1; i >= 0; --i) {
+		auto &info = enqueuedMipsCalls[i];
+		u32 stackRequired = (int)info.args.size() * sizeof(u32) + sizeof(HLEMipsCallStack);
+		u32 stackAligned = ((stackRequired + 15) / 16) * 16;
+
+		sp -= stackAligned;
+		stackData.ptr = sp;
+		stackData->nextOff = stackAligned;
+		stackData->func = info.func;
+		if (info.action) {
+			stackData->actionIndex = (int)mipsCallActions.size();
+			mipsCallActions.push_back(info.action);
+		} else {
+			stackData->actionIndex = 0xFFFFFFFF;
+		}
+		stackData->argc = (int)info.args.size();
+		for (int j = 0; j < (int)info.args.size(); ++j) {
+			Memory::Write_U32(info.args[j], sp + sizeof(HLEMipsCallStack) + j * sizeof(u32));
+		}
+	}
+	enqueuedMipsCalls.clear();
+
+	DEBUG_LOG(HLE, "Executing HLE mips call at %08x, sp=%08x", currentMIPS->pc, sp);
+}
+
+void HLEReturnFromMipsCall() {
+	u32 &sp = currentMIPS->r[MIPS_REG_SP];
+	PSPPointer<HLEMipsCallStack> stackData;
+
+	// At this point, we may have another mips call to run, or be at the end...
+	stackData.ptr = sp;
+
+	if ((stackData->nextOff & 0x0000000F) != 0 || !Memory::IsValidAddress(sp + stackData->nextOff)) {
+		ERROR_LOG(HLE, "Corrupt stack on HLE mips call return: %08x", stackData->nextOff);
+		Core_UpdateState(CORE_ERROR);
+		return;
+	}
+
+	if (stackData->actionIndex != 0xFFFFFFFF && stackData->actionIndex < (u32)mipsCallActions.size()) {
+		PSPAction *&action = mipsCallActions[stackData->actionIndex];
+		VERBOSE_LOG(HLE, "Executing action for HLE mips call at %08x, sp=%08x", stackData->func, sp);
+
+		// Search for the saved v0/v1 values, to preserve the PSPAction API...
+		PSPPointer<HLEMipsCallStack> finalMarker = stackData;
+		while ((finalMarker->nextOff & 0x0000000F) == 0 && Memory::IsValidAddress(finalMarker.ptr + finalMarker->nextOff)) {
```

Comment:
We don't complain here if nextOff is misaligned, maybe should?

Reply:
Here we're just searching (ideally) for nextOff=0xFFFFFFFF.

In the first iteration, it's the same as stackData, so we already complained above if it was invalid.  After that, we would end the loop and we complain afterward if we didn't reach 0xFFFFFFFF (which means that it was misaligned, or otherwise invalid.)

-[Unknown]

Reply:
Oh right, sorry.